exporter/loadbalancingexporter/exporter.go

```diff
@@ -16,42 +16,208 @@ package loadbalancingexporter
 	config Config
+
+	res  resolver
+	ring *hashRing
+
+	exporters            map[string]component.TracesExporter
+	exporterFactory      component.ExporterFactory
+	templateCreateParams component.ExporterCreateParams
+
+	stopped    bool
+	shutdownWg sync.WaitGroup
+	updateLock sync.RWMutex
 }
 
 // Crete new exporter
 func newExporter(params component.ExporterCreateParams, cfg configmodels.Exporter) (*exporterImp, error) {
 	oCfg := cfg.(*Config)
 
+	tmplParams := component.ExporterCreateParams{
+		Logger:               params.Logger,
+		ApplicationStartInfo: params.ApplicationStartInfo,
+	}
+
+	var res resolver
+	if oCfg.Resolver.Static != nil {
+		var err error
+		res, err = newStaticResolver(oCfg.Resolver.Static.Hostnames)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if oCfg.Resolver.DNS != nil { // if both are configured, DNS takes precedence
+		dnsLogger := params.Logger.With(zap.String("resolver", "dns"))
+
+		var err error
+		res, err = newDNSResolver(dnsLogger, oCfg.Resolver.DNS.Hostname)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if res == nil {
+		return nil, errNoResolver
+	}
+
 	return &exporterImp{
 		logger: params.Logger,
 		config: *oCfg,
+
+		res: res,
+
+		exporters:            map[string]component.TracesExporter{},
+		exporterFactory:      otlpexporter.NewFactory(),
+		templateCreateParams: tmplParams,
 	}, nil
 }
 
 func (e *exporterImp) Start(ctx context.Context, host component.Host) error {
+	e.res.onChange(e.onBackendChanges)
+	if err := e.res.start(ctx); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (e *exporterImp) onBackendChanges(resolved []string) {
+	resolved = sort.StringSlice(resolved)
+	newRing := newHashRing(resolved)
+
+	if !newRing.equal(e.ring) {
+		e.updateLock.Lock()
+		defer e.updateLock.Unlock()
+
+		e.ring = newRing
+
+		// TODO: set a timeout?
+		ctx := context.Background()
+
+		// add the missing exporters first
+		e.addMissingExporters(ctx, resolved)
+		e.removeExtraExporters(ctx, resolved)
+	}
+}
+
+func (e *exporterImp) addMissingExporters(ctx context.Context, endpoints []string) {
+	for _, endpoint := range endpoints {
+		endpoint = endpointWithPort(endpoint)
+
+		if _, exists := e.exporters[endpoint]; !exists {
+			cfg := e.exporterFactory.CreateDefaultConfig()
+			oCfg := cfg.(*otlpexporter.Config)
+
+			// is this wrong? are we overriding default fields, or is the config unmarshaller taking care of it?
+			oCfg.ExporterSettings = e.config.Protocol.OTLP.ExporterSettings
+			oCfg.TimeoutSettings = e.config.Protocol.OTLP.TimeoutSettings
+			oCfg.QueueSettings = e.config.Protocol.OTLP.QueueSettings
+			oCfg.RetrySettings = e.config.Protocol.OTLP.RetrySettings
+			oCfg.GRPCClientSettings = e.config.Protocol.OTLP.GRPCClientSettings
+
+			oCfg.Endpoint = endpoint
+			exp, err := e.exporterFactory.CreateTracesExporter(ctx, e.templateCreateParams, cfg)
+			if err != nil {
+				e.logger.Warn("failed to create new trace exporter for endpoint", zap.String("endpoint", endpoint))
```

Comment:
This might be a case where we wouldn't actually update the exporters? What's the implication of the hash-aware load balancing algorithm amidst bad exporters?

Reply:
Great questions. What would happen is:

1. exporters will still be updated with all exporters that were able to be created
1. the hash ring will still have the endpoint for this exporter, but we wouldn't have it in the processor's map, failing at the beginning of the `ConsumeTraces` (`... couldn't find the exporter for the endpoint...`).

I think this is an edge case, as all exporters have the same configuration except for the endpoint, so, creating the trace exporter would fail for all of them already at bootstrap. Endpoints that are offline shouldn't fail at this stage, as the gRPC client won't fail until there's an actual call to the backend, and the retry settings would kick in.

Not sure there's anything of value that we could be making here, but perhaps I'm failing to see what other scenarios could happen.